broker-auth-multiuser-otp/src/main/java/org/aktin/broker/auth/otp/repository/FsUserRepository.java

```diff
@@ -0,0 +1,188 @@
+package org.aktin.broker.auth.otp.repository;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Logger;
+import javax.inject.Singleton;
+
+@Singleton
+public class FsUserRepository implements UserRepository {
+
+  private static final Logger log = Logger.getLogger(FsUserRepository.class.getName());
+
+  private static final String PROPERTY_USER_FILE = "aktin.broker.users.file";
+  private static final String DEFAULT_USER_FILE = "users.txt";
+  private static final String FIELD_SEPARATOR = ";";
+
+  private final Path usersFile;
+  private final Map<String, User> userCache = new ConcurrentHashMap<>();
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
+  private volatile boolean cacheLoaded = false;
+
+  public FsUserRepository() {
+    this(System.getProperty(PROPERTY_USER_FILE, DEFAULT_USER_FILE));
+  }
+
+  public FsUserRepository(String userFilePath) {
+    this.usersFile = Paths.get(userFilePath);
+    initUsersFileIfNonexisting();
+    loadUsersIntoCache();
+  }
+
+  private void initUsersFileIfNonexisting() {
+    if (!Files.exists(usersFile)) {
+      try {
+        if (usersFile.getParent() != null) {
+          Files.createDirectories(usersFile.getParent());
+        }
+        Files.createFile(usersFile);
+        log.info("Created users file: " + usersFile.toAbsolutePath());
+      } catch (IOException e) {
+        log.severe("Could not create users file: " + e.getMessage());
+      }
+    }
+  }
+
+  private void loadUsersIntoCache() {
+    lock.writeLock().lock();
```

Comment:
The locking strategy seems messed up. reading sometimes uses writeLock (loadUsersIntoCache) and sometimes uses readLock (find) and somethimes not locks at all e.g. saveUsersToFile. Also uses ConcurrentHashMap which already handles concurrency for reading so no locking should be needed for accessing the map. loadUsersIntoCache is called in constructor and therefore does not need writeLocking because this cannot coincide with any other call which needs to come after construction. please revise/simplify or discuss with me.